### PR TITLE
doc: Add Memfault user guide

### PR DESCRIPTION
**Pull request by divipillai**
* Add Memfault user guide.
* Removed duplicate content from
 other documentation.
* Provided references to the user guide
from other documentation.

Signed-off-by: divya pillai <divya.pillai@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/ug_memfault.rst [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+1. Add the following Kconfig option in your :file:`prj.conf` file:
+
+   .. code-block:: console
+
+      CONFIG_MEMFAULT=y
+      CONFIG_MEMFAULT_NCS_PROJECT_KEY=""
+
+   The APIs in the Memfault SDK can then be linked into your application.
+   You can get your project key after signing up with Memfault.
+   The :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` option is located in the :file:`../overlay-memfault.conf` file.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:81]:
```suggestion
   The :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option is located in the :file:`../overlay-memfault.conf` file.
```

File: doc/nrf/ug_memfault.rst [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+#. Create a new folder in your project called :file:`config` and add the following three files:
+
+    * :file:`memfault_platform_config.h` - Sets Memfault SDK configurations that are not covered by Kconfig options.
+    * :file:`memfault_metrics_heartbeat_config.def` - Defines application-specific metrics.
+    * :file:`memfault_trace_reason_user_config.def` - Defines application-specific trace reasons.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:87]:
```suggestion
   * :file:`memfault_platform_config.h` - Sets Memfault SDK configurations that are not covered by Kconfig options.
   * :file:`memfault_metrics_heartbeat_config.def` - Defines application-specific metrics.
   * :file:`memfault_trace_reason_user_config.def` - Defines application-specific trace reasons.
```

File: doc/nrf/ug_memfault.rst [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+   Memfault SDK requires these three files in the include path during the build process.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:89]:
```suggestion
   Memfault SDK requires the three files in the include path during the build process.
```

File: doc/nrf/ug_memfault.rst [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+#. Add the following in the :file:`CMakeLists.txt` file to have these configuration files in the include path:
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:91]:
```suggestion
#. Add the following code in the :file:`CMakeLists.txt` file to have these configuration files in the include path:
```

File: doc/nrf/ug_memfault.rst [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+   .. code-block:: console
+
+      zephyr_include_directories(config)
+
+See the `Memfault nRF Connect SDK integration guide`_ documentation for more details.
+
+.. note::
+   In the :ref:`memfault_sample` sample folder, the aforementioned Memfault configurations are already included.
+   You must just add your project key in the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:101]:
```suggestion
   Make sure to add your project key in the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option.
```

File: doc/nrf/ug_memfault.rst [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+You can also add configuration using multiple overlay files, ``-DOVERLAY_CONFIG`` must be set to a list of overlay configurations.
+Following is an example of the overlay details for Memfault integration:
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:104]:
```suggestion
See the following example on how to set an overlay configuration for Memfault integration:
```

File: doc/nrf/ug_memfault.rst [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+.. _ug_memfault_error:
+
+Error Tracking
+==============
+
+Memfault captures all faults that happen at any time.
+The crash data is sent to the Memfault cloud for further inspection and analysis.
+The faults are tracked using the following methods.
+
+Reboot Reason Tracking
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:119]:
```suggestion
Reboot reason tracking
```

File: doc/nrf/ug_memfault.rst [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+----------------------
+
+The device might reboot due to a crash, a brownout, or a firmware update.
+The reboot events for each device are captured and displayed in the Memfault UI.
+
+See `Memfault: Reboot tracking`_ for detailed documentation.
+
+Error Tracking with Trace Events
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:127]:
```suggestion
Error tracking with trace events
```

File: doc/nrf/ug_memfault.rst [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+--------------------------------
+
+Trace event method is used if you want to track errors without capturing full coredump traces (that use less storage) and keeping the device up and running.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:130]:
```suggestion
You can use the trace event method if you want to track errors without capturing full coredump traces (that use less storage) and keep the device up and running.
```

File: doc/nrf/ug_memfault.rst [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+* :ref:`memfault_sample`
+* :ref:`peripheral_mds`
+
+Read the **Error Tracking with trace events** section of these sample documents for more information.
+
+See `Memfault: Error Tracking with Trace Events`_ for detailed documentation.
+Faults are captured and displayed on the **Issues** page of Memfault UI.
+
+RAM-backed stack dump Collection (Coredumps)
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:141]:
```suggestion
RAM-backed stack dump collection (Coredumps)
```

File: doc/nrf/ug_memfault.rst [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+* :ref:`peripheral_mds`
+
+See `Memfault: Collecting Device Metrics`_ for detailed documentation.
+
+Configuration options
+=====================
+
+There are different configuration options defined for the Memfault SDK in |NCS|.
+The configuration options are defined both in `Memfault-SDK`_ and in the integration layer, which integrates the SDK into |NCS|.
+The configuration options of the integration layer are in ``modules/memfault/Kconfig``.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:173]:
```suggestion
The configuration options of the integration layer are in :file:`modules/memfault/Kconfig`.
```

File: doc/nrf/ug_memfault.rst [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+   :alt: Memfault Solution architecture
+
+   Memfault Solution architecture
+
+.. _using_memfault:
+
+Integration overview
+********************
+
+The SDK is part of the West manifest in |NCS| and is automatically downloaded when running ``west update``.
```
Comment (umapraseeda) [inline doc/nrf/ug_memfault.rst:56]:
```suggestion
The SDK version is included in the `west manifest file`_ in |NCS| and is automatically downloaded when running ``west update``.
```

File: doc/nrf/ug_memfault.rst [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+Memfault integration
+####################
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+`Memfault`_ is a cloud based web application with |NCS| SDK compatibility that monitors devices and allows you to debug issues remotely.
+Memfault provides |NCS| with the following features:
```
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/ug_memfault.rst]: After offline discussion with @umapraseeda , rephrased the sentence.

File: doc/nrf/ug_memfault.rst [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+   :local:
+   :depth: 2
+
+
+`Memfault`_ is a cloud-based web application with |NCS| compatibility that monitors devices and allows you to debug issues remotely.
+ Memfault integration provides the following features to |NCS|:
+
+* `Remote debugging <Memfault debugging_>`_ - Memfault issues can be efficiently resolved with automatic detection, alerts, deduplication, and actionable insights sent through the cloud.
+* `Continuous monitoring <Memfault monitoring_>`_ - Monitors device and fleet-level metrics like connectivity and low power state in real-time dashboards.
+  There are 15 prepopulated metrics shown for your devices.
```
Comment (simensrostad) [inline doc/nrf/ug_memfault.rst:16]: Maybe rephrase this to not reference the number of prepoulated metrics (`15`). To reduce maintainance between versions.

File: doc/nrf/ug_memfault.rst [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+1. Add the following Kconfig option in your :file:`prj.conf` file:
+
+   .. code-block:: console
+
+      CONFIG_MEMFAULT=y
+      CONFIG_MEMFAULT_NCS_PROJECT_KEY=""
+
+   The APIs in the Memfault SDK can then be linked into your application.
+   You can get your project key after signing up with Memfault.
+   The :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option is located in the :file:`../overlay-memfault.conf` file.
```
Comment (simensrostad) [inline doc/nrf/ug_memfault.rst:81]: Cannot find any other reference to `overlay-memfault.conf` other than in the ATv2 app. If this is the case this line needs to be removed if the docs are intended to be generic.

File: doc/nrf/ug_memfault.rst [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+#. Create a new folder in your project called :file:`config` and add the following three files:
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:83]: This point is optional if `CONFIG_MEMFAULT_USER_CONFIG_ENABLE` is disabled, and will also not be needed if if it is enabled and `CONFIG_MEMFAULT_USER_CONFIG_SILENT_FAIL` is enabled (this is the default situation).

In short, I think it should be mentioned that this point may be optional depending on configuration.

File: doc/nrf/ug_memfault.rst [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+   * :file:`memfault_platform_config.h` - Sets Memfault SDK configurations that are not covered by Kconfig options.
+   * :file:`memfault_metrics_heartbeat_config.def` - Defines application-specific metrics.
+   * :file:`memfault_trace_reason_user_config.def` - Defines application-specific trace reasons.
+
+   Memfault SDK requires the three files in the include path during the build process.
+
+#. Add the following code in the :file:`CMakeLists.txt` file to have these configuration files in the include path:
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:91]: This point is only needed if the point above is required

File: doc/nrf/ug_memfault.rst [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+You can use the trace event method if you want to track errors without capturing full coredump traces (that use less storage) and keep the device up and running.
+The following samples implement a user-defined trace reason for demonstration purposes.
+
+* :ref:`memfault_sample`
+* :ref:`peripheral_mds`
+
+Read the **Error Tracking with trace events** section of these sample documents for more information.
+
+See `Memfault: Error Tracking with Trace Events`_ for detailed documentation.
+Faults are captured and displayed on the **Issues** page of Memfault UI.
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:139]: It's useful that we are this specific in where to find stuff on Memfault's UI, we just need to be sure to keep it up to date if they changw things

File: doc/nrf/ug_memfault.rst [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+
+See `Memfault: Coredumps`_ for detailed documentation.
+Faults are captured and displayed on the **Issues** page of the Memfault UI.
+
+.. _ug_memfault_metrics:
+
+Collecting Metrics
+==================
+
+The Memfault collects data from each device and visualizes it in a graph.
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:158]: Looks like `SDK` is missing, or if not `The` should probably be removed

File: doc/nrf/ug_memfault.rst [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+
+Applications and samples
+************************
+
+The following application uses the Memfault integration in |NCS|:
+
+* :ref:`asset_tracker_v2` - The debug module of :ref:`asset_tracker_v2` uses `Memfault SDK`_ to track |NCS| specific metrics such as LTE and stack metrics.
+  See :ref:`asset_tracker_v2_debug_module` documentation for more details.
+
+The following sample demonstrates the Memfault integration in |NCS|:
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:187]: `... samples demonstrate ...`

File: doc/nrf/ug_memfault.rst [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+
+* :ref:`peripheral_mds` - This sample has an implementation of the MDS.
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:189]: I'd write `MDS` without abbreviation here, I think

File: doc/nrf/ug_memfault.rst [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+* :ref:`mod_memfault`
+* :ref:`peripheral_mds`
+
+Applications and samples
+************************
+
+The following application uses the Memfault integration in |NCS|:
+
+* :ref:`asset_tracker_v2` - The debug module of :ref:`asset_tracker_v2` uses `Memfault SDK`_ to track |NCS| specific metrics such as LTE and stack metrics.
+  See :ref:`asset_tracker_v2_debug_module` documentation for more details.
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:185]: It could be usefult to mention that the application shows how to use Memfault on a custom transport too, ie not only using Memfault's own HTTPS

Comment (divipillai) [reply doc/nrf/ug_memfault.rst:186]: Had a offline chat with Jan tore, and we agreed to cover this in a separate PR.

File: doc/nrf/ug_memfault.rst (hunk 16 above)
Comment (divipillai) [reply doc/nrf/ug_memfault.rst]: I have re-structured the "Configuring Memfault" section and added a sub-section "Configuration files" that covers the above points. 
Please check.

File: doc/nrf/ug_memfault.rst [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+
+   Memfault Solution architecture
+
+.. _using_memfault:
+
+Integration overview
+********************
+
+The SDK version is included in the `west manifest file`_ in |NCS| and is automatically downloaded when running ``west update``.
+By default, it is downloaded to ``nRF/modules/memfault-firmware-sdk/``.
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:56]: The current path (`nrf/modules/memfault-firmware-sdk/`) is where the NCS integration resides, ie the code that we in Nordic provide to integrate Memfault. 
The Memfault SDK is by default downloaded by west to `ncs/modules/lib/memfault-firmware-sdk/`

File: doc/nrf/ug_memfault.rst [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+
+The APIs in the Memfault SDK can then be linked into your application.
+You can get your project key after signing up with Memfault.
+
+.. note::
+   In the :ref:`memfault_sample` sample folder, the aforementioned Memfault configurations are already included.
+   Make sure to add your project key in the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option.
+
+You can also add configuration using multiple overlay files, ``-DOVERLAY_CONFIG`` must be set to a list of overlay configurations.
+See the following example on how to set an overlay configuration for Memfault integration:
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:84]:
```suggestion
You can also add the configurations using a Kconfig configuration fragment file.
See the following example on how to set an overlay configuration for Memfault integration:
```

Comment (jtguggedal) [reply doc/nrf/ug_memfault.rst:84]: I know we use overlay a lot for these files, but they're now calles Kconfig configuration fragment files according to this one: https://docs.zephyrproject.org/3.2.0/develop/application/index.html#important-build-system-variables

File: doc/nrf/ug_memfault.rst [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+
+Memfault captures all faults that happen at any time.
+The crash data is sent to the Memfault cloud for further inspection and analysis.
+The faults are tracked using the following methods.
+
+Reboot reason tracking
+----------------------
+
+The device might reboot due to a crash, a brownout, or a firmware update.
+The reboot events for each device are captured and displayed in the Memfault UI.
```
Comment (jtguggedal) [inline doc/nrf/ug_memfault.rst:124]:
```suggestion
The reboot events for each device are captured and displayed in the Memfault UI along with the recorded reboot reason.
```

**Review by simensrostad — COMMENTED** [review]
lgtm

**Review by simensrostad — APPROVED** [review]

**Review by jtguggedal — APPROVED** [review]

**Review by umapraseeda — APPROVED** [review]
